Sources/Basics/VirtualFileSystem.swift

```diff
@@ -0,0 +1,225 @@
+    }
+
+    public func isFile(_ path: AbsolutePath) -> Bool {
+        return findNode(path, followSymlink: true)?.isFile == true
+    }
+
+    public func isExecutableFile(_ path: AbsolutePath) -> Bool {
+        guard let node = findNode(path, followSymlink: true) else { return false }
+        if case let .file(_, isExecutable, _) = node {
+            return isExecutable
+        } else {
+            return false
+        }
+    }
+
+    public func isSymlink(_ path: AbsolutePath) -> Bool {
+        return findNode(path, followSymlink: true)?.isSymlink == true
+    }
+
+    public func getDirectoryContents(_ path: AbsolutePath) throws -> [String] {
+        guard let node = findNode(path, followSymlink: true) else { throw Errors.noSuchFileOrDirectory(path: path) }
+        return node.children.map { $0.name }
+    }
+
+    public var currentWorkingDirectory: AbsolutePath? = nil
+
+    public func changeCurrentWorkingDirectory(to path: AbsolutePath) throws {
+        throw Errors.readOnlyFileSystem
+    }
+
+    public var homeDirectory = AbsolutePath("/")
+
+    public var cachesDirectory: AbsolutePath? = nil
+
+    public func createSymbolicLink(_ path: AbsolutePath, pointingAt destination: AbsolutePath, relative: Bool) throws {
+        throw Errors.readOnlyFileSystem
+    }
+
+    public func removeFileTree(_ path: AbsolutePath) throws {
+        throw Errors.readOnlyFileSystem
+    }
+
+    public func copy(from sourcePath: AbsolutePath, to destinationPath: AbsolutePath) throws {
+        throw Errors.readOnlyFileSystem
+    }
+
+    public func move(from sourcePath: AbsolutePath, to destinationPath: AbsolutePath) throws {
+        throw Errors.readOnlyFileSystem
+    }
+
+    public func createDirectory(_ path: AbsolutePath, recursive: Bool) throws {
+        throw Errors.readOnlyFileSystem
+    }
+
+    public func readFileContents(_ path: AbsolutePath) throws -> ByteString {
+        guard let node = findNode(path, followSymlink: true) else { throw Errors.noSuchFileOrDirectory(path: path) }
+        if !node.isFile {
+            throw Errors.notAFile(path: path)
+        }
+        return try readFileContents(path)
+    }
+
+    public func writeFileContents(_ path: AbsolutePath, bytes: ByteString) throws {
+        throw Errors.readOnlyFileSystem
+    }
+
+    public func chmod(_ mode: FileMode, path: AbsolutePath, options: Set<FileMode.Option>) throws {
+        throw Errors.readOnlyFileSystem
+    }
+
+    public func getFileInfo(_ path: AbsolutePath) throws -> FileInfo {
+        guard let node = findNode(path, followSymlink: true) else { throw Errors.noSuchFileOrDirectory(path: path) }
+
+        let attrs: [FileAttributeKey: Any] = [
+            .systemNumber: NSNumber(value: UInt64(0)),
+            .systemFileNumber: UInt64(0),
+            .posixPermissions: NSNumber(value: Int16(0)),
+            .type: node.fileAttributeType,
+            .size: UInt64(0),
```

Comment:
Shouldn't this (and the permissions, etc) be part of the input JSON?

Reply:
Eventually yes, I think this is good enough for an initial version of this.